c-scape/src/fs/open.rs

```diff
@@ -0,0 +1,77 @@
+use rustix::fd::{BorrowedFd, IntoRawFd};
+use rustix::ffi::ZStr;
+use rustix::fs::{cwd, Mode, OFlags};
+
+use libc::{c_char, c_int};
+
+use crate::convert_res;
+
+macro_rules! openat_impl {
+    ($fd: expr, $pathname: ident, $flags: ident, $args: ident) => {{
+        let flags = OFlags::from_bits($flags as _).unwrap();
+        let mode = if flags.contains(OFlags::CREATE) {
+            let mode: libc::mode_t = $args.arg();
+            Mode::from_bits((mode & !libc::S_IFMT) as _).unwrap()
+        } else {
+            Mode::empty()
+        };
+        match convert_res(rustix::fs::openat(
+            &$fd,
+            ZStr::from_ptr($pathname.cast()),
+            flags,
+            mode,
+        )) {
+            Some(fd) => fd.into_raw_fd(),
+            None => -1,
+        }
+    }};
+}
+
+// we open all files with O_LARGEFILE as that is what Rustix does
+// hopefully this doesn't break any C programs
+#[no_mangle]
+unsafe extern "C" fn open(pathname: *const c_char, flags: c_int, mut args: ...) -> c_int {
+    libc!(libc::open(pathname, flags, args));
+
+    openat_impl!(cwd(), pathname, flags, args)
```

Comment:
Is there a reason for using `openat_impl` instead of just calling open here? I noticed for other functions, like lseek, the non-64 version is just a wrapper around the 64 version.

Reply:
The thing is these are varadic functions, which can't be delegated (the libc macro above actually can't be called without being UB), so we have to use a macro. I'll probably figure out how to write this to use VaList / move some parts out of the macro and delegate normally.

Reply:
Ah, makes sense. Either way sounds reasonable, but please add a comment calling this out.